Wrappers/Python/cil/optimisation/functions/SubsetGradientAlgorithmFunction.py

```diff
@@ -0,0 +1,122 @@
+from cil.optimisation.functions import Function, SumFunction
+import numpy as np
+
+class SubsetGradientAlgorithmFunction(SumFunction):
+    
+    '''Class for use as objective function in gradient type algorithms to enable the use of subsets.
+
+    The `gradient` method is implemented in children classes and allows to return an approximation of the gradient based on subset gradients.
+
+    Parameters:
+    -----------
+
+
+
+    '''
+    
+    # name suggestions: CustomGradientFunction, 
+    #                   A name with the subsets word
+    #     
+
+    
+    def __init__(self, functions, **kwargs):
+        # should not have docstring
+        super(SubsetGradientAlgorithmFunction, self).__init__(*functions)
+        
+    @property
+    def num_subsets(self):
+        return len(self.functions)
+        
+    def _full_gradient(self, x, out=None):
+        '''Return full gradient'''
+        return super(SubsetGradientAlgorithmFunction, self).gradient(x, out=out)
+        
+    def gradient(self, x, out=None):
+        """        
+            maybe calls _approx_grad inside        
+        """
+        raise NotImplemented
+
+
+    def subset_gradient(self, x, subset_num,  out=None):
+
+        return self.functions[subset_num].gradient(x)
+
+
+    def next_subset(self):
+
+        raise NotImplemented
+
+
+class SAGA_Function(SubsetGradientAlgorithmFunction):
```

Comment:
This should be renamed to `SAGAGradientFunction`

Reply:
done in 41fb62aa3f5bc835b4651aa562d809081ff1df64